### PR TITLE
logging: Added BLE Backend Service

### PR DESCRIPTION
**Pull request by vChavezB**
## Problem

There is not a logger to be used over a wireless connection for tests and mantainance. Similar issues

https://github.com/zephyrproject-rtos/zephyr/issues/12714
https://github.com/zephyrproject-rtos/zephyr/issues/34949

## Summary
Added a new logger backend to recieve data over a BLE notification characteristic. The characteristic is based on the UUID from the Nordic connect SDK service NUS,  which allows to have a UART shell over BLE. 

The idea behind this, is that this logger can be used directly with the NRF apps or any other BLE UART terminal app.  

Feel free to discuss and give any feedback.

### CODE REVIEW

**Comment by jori-nordic** [comment]
Thanks! I'll try it out this week. Could you add a sample that enables it? That way you can also add more documentation on how it works (ie, what happens to the logs when the device is disconnected or unsubscribes, when it it reconnected, etc).

**Comment by vChavezB** [comment]
Thanks for the suggestion. Where should I put the example ? 

`samples\subsys\logging\ble_backend` ?

**Comment by vChavezB** [comment]
I updated the commit with a sample application. Some more documentation is required but this way you can test the ble backend at your end.

File: subsys/logging/backends/Kconfig.ble [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+	  notification characteristic. The minimum size is 20 for the smallest
+	  MTU packet. Be sure to increase the MTU size in your application to use
+	  bigger values. See BT_L2CAP_TX_MTU,BT_BUF_ACL_TX_SIZE,BT_CTLR_DATA_LENGTH_MAX
+
+
+backend = BLE
+backend-str = ble
+source "subsys/logging/Kconfig.template.log_format_config"
+
+endif # LOG_BACKEND_SWO
```
Comment (jori-nordic) [inline subsys/logging/backends/Kconfig.ble:35]: update

**Review by jori-nordic — CHANGES_REQUESTED** [review]
Thanks for the patch!
I have tested it with both nRF connect & nRF toolbox apps, works a treat.

It does block the whole OS if the buffer size is misconfigured (e.g. large buffer size, but the log sink/phone app doesn't do the necessary MTU update), but I think that it's fine enough for experimental use.

I do have some suggestions below

File: samples/subsys/logging/ble_backend/sample.yaml [hunk 2]
```diff
@@ -0,0 +1,4 @@
+sample:
+  description: A simple application that demonstrate
+    use of the logging ble backend.
+  name: logger backend ble
```
[the remark this answers is not in the document]
Comment (vChavezB) [reply samples/subsys/logging/ble_backend/sample.yaml:4]: OK I will add this. I did not know how the CI is integrated so I left this file as simple as possible for further feedback

**Comment by vChavezB** [comment]
I have updated the PR with the proposed changes

**Review by jori-nordic — DISMISSED** [review]
Just needs a formatting pass, take a look at the warnings (they show up in the "files changed" tab)

**Comment by jori-nordic** [comment]
@stephanosio Could you help us out? We're trying to add this sample to the CI builds, but I've never done this before.
I tried copying from one of the sample yamls, but there's still an error.

**Comment by vChavezB** [comment]
As I do not know how Twister and its integration on the pipeline I decided to use to use instead the nrf52833 dk as platform for the test.I t looks like it works. 

Nevertheless if its important to do the tests with the suggested sample.yaml

```yaml
  platform_allow: qemu_cortex_m3 qemu_x86
  integration_platforms:
    - qemu_cortex_m3
```
then the issue should be checked: https://github.com/zephyrproject-rtos/zephyr/actions/runs/4574368071/jobs/8075903426

**Comment by jori-nordic** [comment]
I think the original should work. This is what seems to be breaking the CI pipeline - seems unrelated - could you revert the changes back to the original integration/test platform?
image

**Comment by vChavezB** [comment]
> I think the original should work. This is what seems to be breaking the CI pipeline - seems unrelated - could you revert the changes back to the original integration/test platform? image

reverted

**Comment by vChavezB** [comment]
Fixed compliance checks

**Review by theob-pro — CHANGES_REQUESTED** [review]

File: samples/subsys/logging/ble_backend/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,11 @@
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.20.0)
+
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+project(peripheral_hr)
```
[the remark this answers is not in the document]
Comment (vChavezB) [reply samples/subsys/logging/ble_backend/CMakeLists.txt:6]: thanks for catching that. I will change it

File: samples/subsys/logging/ble_backend/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+
+FILE(GLOB app_sources src/*.c)
+target_sources(app PRIVATE
+  ${app_sources}
+  )
```
[the remark this answers is not in the document]
Comment (vChavezB) [reply samples/subsys/logging/ble_backend/CMakeLists.txt:11]: ah yes, there is no need to use global here. Will change it. I took the CMakeLists.txt from another example. My bad!

File: subsys/logging/backends/CMakeLists.txt [hunk 5]
```diff
@@ -15,6 +15,12 @@ zephyr_sources_ifdef(
   log_backend_adsp_mtrace.c
 )
 
+zephyr_sources_ifdef(
+  CONFIG_LOG_BACKEND_BLE
+  log_backend_ble.c
+)
+
+
 zephyr_sources_ifdef(
```
Comment (theob-pro) [inline subsys/logging/backends/CMakeLists.txt:24]:
```suggestion
zephyr_sources_ifdef(
  CONFIG_LOG_BACKEND_BLE
  log_backend_ble.c
)

zephyr_sources_ifdef(
```
Small change to be consistent in the file.

**Comment by vChavezB** [comment]
Added changes according to comments from @theob-pro

**Review by jori-nordic — APPROVED** [review]
tested again after the prj.conf changes, still works :+1:

**Review by theob-pro — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Review by nordic-krch — APPROVED** [review]

**Comment by fabiobaltieri** [comment]
> Hi @fabiobaltieri, can you open an issue so I can try to trouble shoot the second point about only enabling the BLE Backend.

Sure thing.

**Comment by fabiobaltieri** [comment]
> The nRF toolbox app does it transparently (if the host supports it) so it just works. If you use the nRF connect app, you need to do it manually (there's an option somewhere). And ofc if you're using a zephyr sample as the sink, then you'll need to do it manually as well.

Got it, yeah I'm using nRF Connect on a Pixel 6 and it does indeed work if I manually trigger an MTU update. The problem is that when it fails it seems to be spinning somewhere (https://gist.github.com/fabiobaltieri/e062b27f289378a2b3ae3da666940ffb) and it blocks the application, it may as well be crashing it. It should probably discard the log and continue instead.

**Pull request merged**